modules/basegl/src/processors/shadercomponentprocessorbase.cpp

```diff
@@ -0,0 +1,138 @@
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <modules/basegl/processors/shadercomponentprocessorbase.h>
+
+#include <inviwo/core/util/rendercontext.h>
+
+#include <modules/opengl/volume/volumegl.h>
+#include <modules/opengl/texture/textureunit.h>
+#include <modules/opengl/texture/textureutils.h>
+#include <modules/opengl/shader/shaderutils.h>
+#include <modules/opengl/shader/shadermanager.h>
+
+#include <modules/basegl/shadercomponents/shadercomponent.h>
+
+namespace inviwo {
+
+ShaderComponentProcessorBase::ShaderComponentProcessorBase(
+    const std::vector<std::pair<ShaderType, std::shared_ptr<const ShaderResource>>>&
+        shaderResources,
+    std::string_view identifier, std::string_view displayName)
+    : Processor(identifier, displayName)
+    , outport_("outport")
+    , shader_{shaderResources, Shader::Build::No}
+    , components_{} {
+
+    shader_.onReload([this]() { invalidate(InvalidationLevel::InvalidResources); });
+
+    addPort(outport_, "images");
+}
+
+ShaderComponentProcessorBase::~ShaderComponentProcessorBase() = default;
+
+void ShaderComponentProcessorBase::registerComponent(ShaderComponent& component) {
+    components_.push_back(&component);
+    for (const auto& [port, group] : component.getInports()) {
+        addPort(*port, group);
+    }
+    for (auto prop : component.getProperties()) {
+        addProperty(*prop);
+    }
+}
+
+void ShaderComponentProcessorBase::initializeResources() {
+    for (auto* comp : components_) {
+        try {
+            comp->initializeResources(shader_);
+        } catch (...) {
+            handleError("initializeResources", comp->getName());
+        }
+    }
+
+    auto* fso = shader_.getFragmentShaderObject();
+    fso->clearOutDeclarations();
+
+    fso->clearSegments();
+    for (auto* comp : components_) {
+        try {
+            for (auto&& segment : comp->getSegments()) {
+                fso->addSegment(ShaderSegment{segment.placeholder, std::string(comp->getName()),
+                                              segment.snippet, segment.priority});
+            }
+        } catch (...) {
+            handleError("adding shader segments", comp->getName());
+        }
+    }
+
+    shader_.build();
+}
+
+void ShaderComponentProcessorBase::process() {
+    utilgl::activateAndClearTarget(outport_);
+    shader_.activate();
+
+    TextureUnitContainer units;
+    utilgl::setUniforms(shader_, outport_);
+
+    for (auto& comp : components_) {
+        try {
+            comp->process(shader_, units);
+        } catch (...) {
+            handleError("processing", comp->getName());
+        }
+    }
+
+    utilgl::singleDrawImagePlaneRect();
```

Comment:
That is very volume rendering specific. Is this mentioned in the docs that a full-screen quad is rendered?

Reply:
should be easy to move to a virtual(abstract) draw function or something like that. But I rather do that when I see the need, then before hand.